### PR TITLE
BUG: Fix buffer export dtype references

### PR DESCRIPTION
**Pull request by seberg**
There is a small refcount bug introduced when fixing/improving the
buffer export code recently. This fixes it.

---

It would be fine to just merge this, I am running the test more, so I might find more small issues like this, but it might also take another day until I even look/get the test results...

### CODE REVIEW

**Comment by seberg** [comment]
This needs to be backported if the other buffer fixes get backported.

File: numpy/core/src/multiarray/buffer.c [hunk 1]
```diff
@@ -544,7 +547,6 @@ _buffer_info_new(PyObject *obj, int flags)
                 }
             }
         }
-        Py_INCREF(descr);
     }
 
     /* Fill in format */
```
Comment (mattip) [inline numpy/core/src/multiarray/buffer.c:552]: `descr` is not used until line 554. Could you move the `descr = PyArray_DESCR(arr); Py_INCREF(descr);` to there instead, or comment why it needs to stay at line 489?

**Comment by mattip** [comment]
Could you xref the other PRs that this relates to, so when/if backporting them this has less chance of being forgotten?

File: numpy/core/src/multiarray/buffer.c (hunk 1 above)
Comment (seberg) [reply numpy/core/src/multiarray/buffer.c]: Yeah, much better, thanks, definitely moved the wrong line around (moved it above now as well, so its at the end of the block in both branches).

**Comment by seberg** [comment]
The PR where this was introduced is gh-17295.

**Review by mattip — APPROVED** [review]

**Comment by charris** [comment]
> This needs to be backported if the other buffer fixes get backported.

I'm planning on a 1.19.5 for the segfault fix, and maybe an OpenBLAS update if that gets sorted. 1.20.0 is close enough that I think lesser fixes can wait. How much importance do you put on the buffer fixes?

**Comment by seberg** [comment]
I assume you didn't merge because the test run didn't finish. It seems to be handing, so I will just ignore it. Only the refcheck tests were missing it appears, those are really unlikely to be related to anything here.

Thanks Matti.

**Pull request merged**